### PR TITLE
[#10176] Fix statistics-related problems in session results page

### PR DESCRIPTION
**Pull request by wkurniawan07**
Part of #10176 

This fixes the following tickets:
- Numerical scale statistics duplicated when collapsing and expanding the result tab
- GRQ, RGQ, GQR, RQG alice betsy > alice betsy, q14 missing (also observed in V6)
- GRQ/RGQ/GQR/RQG view: Teams displayed in sections they don't belong to
- In GRQ/RGQ view, the show statistics option should be disable as no statistics are shown
- GQR/RQG view: missing team statistics
- Question View > Any section > statistics table do not update to show the statistics of the filtered responses
- No response behavior different (V7 shows result with no response, V6 omits them)
- In student all records page, if there are no responses for a student, V6 displays a No feedback for xxx found. But in V7, all tabs are shown (same issue as the previous item)

### CODE REVIEW

File: src/web/app/components/question-responses/grouped-responses/grouped-responses.component.ts [hunk 1]
```diff
@@ -42,11 +42,24 @@ export class GroupedResponsesComponent extends InstructorResponsesViewBase imple
...
   };
 
+  hasNonMissingResponses: boolean = false;
+
   constructor() {
     super();
   }
 
   ngOnInit(): void {
+    for (const question of this.responses) {
```
Comment (xpdavid) [inline src/web/app/components/question-responses/grouped-responses/grouped-responses.component.ts:52]: - `this.hasNonMissingResponses ` -> `this.hasRealResponses` (or we can just go opposite `this.areAllMissingResponses`). Reasons: The name `hasNonMissingResponses` seems it requires non missing response while it is not true in the case.

- And to save code :P
```javascript
this.hasNonMissingResponses = this.responses.some((question) => question.allResponses.some((response) => !response.isMissingResponse))
```

File: src/web/app/components/question-responses/grq-rgq-view-responses/grq-rgq-view-responses.component.ts [hunk 2]
```diff
@@ -56,6 +56,7 @@ export class GrqRgqViewResponsesComponent extends InstructorResponsesViewBase im
   userExpanded: Record<string, boolean> = {};
 
   responsesToShow: Record<string, Record<string, QuestionOutput[]>> = {};
+  userHasNonMissingResponses: Record<string, boolean> = {};
```
Comment (xpdavid) [inline src/web/app/components/question-responses/grq-rgq-view-responses/grq-rgq-view-responses.component.ts:59]: Same naming issue here.

File: src/web/app/components/question-responses/grq-rgq-view-responses/grq-rgq-view-responses.component.ts [hunk 3]
```diff
@@ -174,6 +178,14 @@ export class GrqRgqViewResponsesComponent extends InstructorResponsesViewBase im
             this.responsesToShow[user] = this.responsesToShow[user] || {};
             this.responsesToShow[user][other] = this.responsesToShow[user][other] || [];
             this.responsesToShow[user][other].push(questionCopy2);
+
+            if (!this.userHasNonMissingResponses[user]) {
+              const hasNonMissingResponse: boolean =
+                  !!questionCopy2.allResponses.filter((response: ResponseOutput) => !response.isMissingResponse).length;
```
Comment (xpdavid) [inline src/web/app/components/question-responses/grq-rgq-view-responses/grq-rgq-view-responses.component.ts:184]:
```javascript
this.userHasNonMissingResponses[user] = questionCopy2.allResponses.some((response: ResponseOutput) => !response.isMissingResponse)
```

File: src/web/app/components/question-responses/per-question-view-responses/per-question-view-responses.component.ts [hunk 4]
```diff
@@ -89,7 +89,14 @@ export class PerQuestionViewResponsesComponent extends InstructorResponsesViewBa
...
       responsesToShow.push(response);
     }
-    this.responsesToShow = responsesToShow;
+
+    const hasNonMissingResponse: boolean =
+        !!responsesToShow.filter((response: ResponseOutput) => !response.isMissingResponse).length;
+    if (hasNonMissingResponse) {
+      this.responsesToShow = responsesToShow;
+    } else {
+      this.responsesToShow = [];
```
Comment (xpdavid) [inline src/web/app/components/question-responses/per-question-view-responses/per-question-view-responses.component.ts:98]: So we are saying if all responses are missing responses then we are going to show it as no response? Looks confusing under the case where `indicateMissingResponses` is true.

Comment (wkurniawan07) [reply src/web/app/components/question-responses/per-question-view-responses/per-question-view-responses.component.ts:98]: Knowing which responses are missing is not meaningful when there are no actual responses that act as the control.

File: src/web/app/components/question-responses/gqr-rqg-view-responses/gqr-rqg-view-responses.component.ts [hunk 5]
```diff
@@ -86,19 +86,26 @@ export class GqrRqgViewResponsesComponent extends InstructorResponsesViewBase im
...
         }
-        if (response.giverEmail) {
-          this.userToEmail[response.giver] = response.giverEmail;
-        }
-        if (response.recipientEmail) {
-          this.userToEmail[response.recipient] = response.recipientEmail;
-        }
 
+        if (this.section) {
+          if (this.isGqr && response.giverSection !== this.section
```
[the remark this answers is not in the document]
Comment (wkurniawan07) [reply src/web/app/components/question-responses/gqr-rqg-view-responses/gqr-rqg-view-responses.component.ts:93]: Redundant, not invalid.

[the remark this answers is not in the document]
Comment (xpdavid) [reply src/web/app/components/question-responses/gqr-rqg-view-responses/gqr-rqg-view-responses.component.ts:93]: I think somehow later we should just hide the options. In addition, I think if we do this properly, we might not need this if check and the `isFeedbackResponsesDisplayedOnSection` can filter responses already.

File: src/web/app/pages-instructor/instructor-session-result-page/instructor-session-result-page.component.html [hunk 6]
```diff
@@ -109,7 +109,7 @@ <h1>
         <label class="form-check-label" ngbTooltip="Group results in the current view by team"><input type="checkbox" [(ngModel)]="groupByTeam" [disabled]="viewType === 'QUESTION'"> Group by Teams</label>
       </div>
       <div>
-        <label class="form-check-label"  ngbTooltip="Show statistics"><input type="checkbox" [(ngModel)]="showStatistics"> Show Statistics</label>
+        <label class="form-check-label" ngbTooltip="Show statistics"><input type="checkbox" [(ngModel)]="showStatistics" [disabled]="viewType === 'GRQ' || viewType === 'RGQ'"> Show Statistics</label>
```
[the remark this answers is not in the document]
Comment (xpdavid) [reply src/web/app/pages-instructor/instructor-session-result-page/instructor-session-result-page.component.html:112]: Are we going to address this?

> The checkbox should also be unchecked (to show that stats is not enabled and not allowed to enabled).

File: src/web/app/components/question-responses/grq-rgq-view-responses/grq-rgq-view-responses.component.ts [hunk 7]
```diff
@@ -174,6 +178,14 @@ export class GrqRgqViewResponsesComponent extends InstructorResponsesViewBase im
             this.responsesToShow[user] = this.responsesToShow[user] || {};
             this.responsesToShow[user][other] = this.responsesToShow[user][other] || [];
             this.responsesToShow[user][other].push(questionCopy2);
+
+            if (!this.userHasRealResponses[user]) {
+              const hasRealResponse: boolean =
```
Comment (xpdavid) [inline src/web/app/components/question-responses/grq-rgq-view-responses/grq-rgq-view-responses.component.ts:183]:
```suggestion
              this.userHasRealResponses[user] =
```

can be simplified :P

File: src/web/app/pages-instructor/instructor-session-result-page/instructor-session-result-page.component.html (hunk 6 above)
[the remark this answers is not in the document]
Comment (wkurniawan07) [reply src/web/app/pages-instructor/instructor-session-result-page/instructor-session-result-page.component.html:112]: Don't think it's worth the trouble.

**Pull request merged**